src/cargo/ops/vendor.rs

```diff
@@ -237,12 +249,47 @@ fn sync(
         paths::write(&cksum, json.to_string())?;
     }
 
-    for path in to_remove {
-        if path.is_dir() {
-            paths::remove_dir_all(&path)?;
+    let old_sources: BTreeSet<String> = if sources_file.exists() {
+        // Read from `<vendor-dir>/.sources` file
+        let bytes = paths::read_bytes(sources_file)?;
+        serde_json::from_slice(&bytes)?
+    } else {
+        BTreeSet::new()
+    };
+
+    if !opts.no_delete {
+        if contains_duplicate_versions {
+            let old_source_paths = old_sources.iter().map(|p| canonical_destination.join(p));
+            for path in old_source_paths {
+                remove_dir_contents(&path, &vendored_crates)?;
+            }
         } else {
-            paths::remove_file(&path)?;
+            remove_dir_contents(canonical_destination, &vendored_crates)?;
+        };
+    }
+
+    if contains_duplicate_versions {
+        // Write the sources list back to `<vendor-dir>/.sources` file.
+        let mut new_sources: BTreeSet<_> = sources
+            .iter()
+            .map(|src_id| source_id_to_dir_name(*src_id))
+            .collect();
+        let diff = old_sources
+            .difference(&new_sources)
+            .cloned()
+            .collect::<Vec<_>>();
+        for dir_name in diff {
+            let path = canonical_destination.join(&dir_name);
+            if path.is_dir() {
+                if path.read_dir()?.next().is_none() {
+                    // Cleanup empty directories
+                    paths::remove_dir(path)?;
+                } else {
+                    new_sources.insert(dir_name);
+                }
+            }
         }
+        paths::write(sources_file, serde_json::to_string(&new_sources)?)?;
```

Comment:
I think it would ideally be best if we could avoid having this file altogether, but I might be missing something. Could you explain a bit more why this file is needed and what it's used for?

Reply:
Sorry for not think more on it. This logic is also from `cargo-vendor`, specifically from this PR: https://github.com/alexcrichton/cargo-vendor/pull/96. IIUC this file is served to identify whether or not a folder is path registry previously vendored. I have proposed a solution in https://github.com/rust-lang/cargo/pull/10344#discussion_r798154686 and hope it help to avoid the file.